qiskit/aqua/algorithms/amplitude_amplifiers/grover.py

```diff
@@ -111,192 +128,274 @@ def __init__(self,
+            statevector = result.get_statevector(qc)
+            num_bits = len(self._grover_operator.reflection_qubits)
+            # trace out work qubits
+            if qc.width() != num_bits:
+                rho = get_subsystem_density_matrix(
+                    statevector,
+                    range(num_bits, qc.width())
+                )
+                statevector = np.diag(rho)
+            max_amplitude = max(statevector.max(), statevector.min(), key=abs)
+            max_amplitude_idx = np.where(statevector == max_amplitude)[0][0]
+            top_measurement = np.binary_repr(max_amplitude_idx, num_bits)
+
         else:
-            qc = self.construct_circuit(measurement=True)
+            qc = self.construct_circuit(power, measurement=True)
             measurement = self._quantum_instance.execute(qc).get_counts(qc)
             self._ret['measurement'] = measurement
             top_measurement = max(measurement.items(), key=operator.itemgetter(1))[0]
 
         self._ret['top_measurement'] = top_measurement
-        oracle_evaluation, assignment = self._oracle.evaluate_classically(top_measurement)
-        return assignment, oracle_evaluation
 
-    def construct_circuit(self, measurement=False):
-        """
-        Construct the quantum circuit
+        return self.post_processing(top_measurement), self.is_good_state(top_measurement)
+
+    def is_good_state(self, bitstr: str) -> bool:
+        """Check whether a provided bitstring is a good state or not.
 
         Args:
-            measurement (bool): Boolean flag to indicate if
-                measurement should be included in the circuit.
+            bitstr: The measurement as bitstring.
+
+        Raises:
+            NotImplementedError: If self._is_good_state couldn't be used to determine whether
+                the bitstring is a good state.
 
         Returns:
-            QuantumCircuit: the QuantumCircuit object for the constructed circuit
+            True if the measurement is a good state, False otherwise.
         """
-        if self._incremental:
-            if self._qc_amplitude_amplification is None:
-                self._qc_amplitude_amplification = \
-                    QuantumCircuit() + self.qc_amplitude_amplification_iteration
+        if callable(self._is_good_state):
+            oracle_evaluation, _ = self._is_good_state(bitstr)
+            return oracle_evaluation
+        elif isinstance(self._is_good_state, list):
+            return bitstr in self._is_good_state
+        elif isinstance(self._is_good_state, Statevector):
+            return bitstr in self._is_good_state.probabilities_dict()
         else:
-            self._qc_amplitude_amplification = QuantumCircuit()
-            for _ in range(self._num_iterations):
-                self._qc_amplitude_amplification += self.qc_amplitude_amplification_iteration
+            raise NotImplementedError('Conversion to callable not implemented for {}'.format(
```

Comment:
This type check, if you feel its needed to be done, would be better in the constructor where this good state is passed - checking it here seems late to me given that we can do it when this is built and not wait until things are run

Reply:
I would say it's not needed since we can expect users to obey the type hints. I'll update it.